### PR TITLE
interactor: introduce zeroconf

### PR DESCRIPTION
**Pull request by delfick**

### CODE REVIEW

**Comment by Djelibeybi** [comment]
While playing with something else, Interactor decided to stop playing after reporting this warning:
```
14:59:35 WARNING zeroconf        Received invalid packet from ('192.168.254.1', 5351) at offset 52819 while unpacking b'\x02\x81\x00\x00\x00\x00\x1c \x00\x07F\xba\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xb4\xdd\xd9ZD\xe7\xd0^\xce1\x10\xab\x11\x00\x00\x00\x14\xe9\x14\xf0\x00\x00\x00\x00\x00\x00\x00\x00\x00\x00\xff\xffg3r\xa3'
Traceback (most recent call last):
  File "/Users/amiller/Git/Djelibeybi/photons/venv/lib/python3.9/site-packages/zeroconf/_protocol/incoming.py", line 111, in _parse_data
    parser_call()
  File "/Users/amiller/Git/Djelibeybi/photons/venv/lib/python3.9/site-packages/zeroconf/_protocol/incoming.py", line 105, in _initial_parse
    self.read_others()
  File "/Users/amiller/Git/Djelibeybi/photons/venv/lib/python3.9/site-packages/zeroconf/_protocol/incoming.py", line 185, in read_others
    domain = self.read_name()
  File "/Users/amiller/Git/Djelibeybi/photons/venv/lib/python3.9/site-packages/zeroconf/_protocol/incoming.py", line 275, in read_name
    self.offset = self._decode_labels_at_offset(self.offset, labels, seen_pointers)
  File "/Users/amiller/Git/Djelibeybi/photons/venv/lib/python3.9/site-packages/zeroconf/_protocol/incoming.py", line 315, in _decode_labels_at_offset
    raise IncomingDecodeError("Corrupt packet received while decoding name")
zeroconf._exceptions.IncomingDecodeError: Corrupt packet received while decoding name
```

**Comment by delfick** [comment]
weird. zeroconf bug?

**Comment by Djelibeybi** [comment]
Dunno. I just saw it, so I figured I'd log it for one of us to look at later. I'm busy working on the thing that needs this thing, so I'll likely have a poke at it at some point.

**Comment by Djelibeybi** [comment]
It looks like it's either a ZeroConf or UniFi mDNS reflector issue. Will look deeper later.

**Comment by Djelibeybi** [comment]
Updated to match RFC and other implementation standards after spending way too long staring at the output of `dns-sd`.

File: apps/interactor/interactor/zeroconf.py [hunk 1]
```diff
@@ -47,30 +50,27 @@ async def finish(self, exc_typ=None, exc=None, tb=None):
...
-                if netifaces.AF_INET not in addresses:
-                    continue
-
-                for address in addresses[netifaces.AF_INET]:
-                    if address["addr"] != "127.0.0.1":
-                        val = address["addr"]
-
-        if val is sb.NotSpecified:
-            raise BadSpecValue("Failed to figure out an IP address to use")
+            """Get the currently active/routing IP address of the local machine."""
```
Comment (delfick) [inline apps/interactor/interactor/zeroconf.py:65]: I'm not a fan of this way of doing this. How is it better?

File: apps/interactor/interactor/zeroconf.py [hunk 2]
```diff
@@ -1,33 +1,36 @@
...
     async def start(self):
-        log.info(
-            hp.lc(
-                "Enabling service discovery",
-                address=f"{self.options.ip_address}:{self.port}",
-                dns=f"{self.options.interactor_name}.{self.options.service_type_name}",
-            )
-        )
         if self.options.enabled:
+            log.info(
```
Comment (delfick) [inline apps/interactor/interactor/zeroconf.py:25]: oops lol

**Review by delfick — COMMENTED** [review]
nice, thanks

File: apps/interactor/interactor/zeroconf.py (hunk 1 above)
Comment (Djelibeybi) [reply apps/interactor/interactor/zeroconf.py:65]: This method finds the device that is used to route outound via the highest priority gateway on the box. It's the better way of reducing to a single, most likely to be operational, interface.

Alternatively, we could just list all interfaces (inclding 127.0.0.1). There are a few devices on my network that appear to do that.

**Comment by Djelibeybi** [comment]
> weird. zeroconf bug?

Known issue: https://github.com/jstasiak/python-zeroconf/issues/993

File: apps/interactor/interactor/zeroconf.py (hunk 1 above)
Comment (delfick) [reply apps/interactor/interactor/zeroconf.py:65]: > most likely to be operational, interface.

I suppose.

My way has a better opportunity for which interface to use to be configurable though.

I've decided I don't have a strong enough opinion to care either way, this way does seem to work on my laptop *shrugs*

Comment (Djelibeybi) [reply apps/interactor/interactor/zeroconf.py:65]: We need to import socket for other things so this reduces the import count by one.

But it's also a hill upon which I'm not even prepared to get slightly inconvenienced so I'm happy to be overruled.

Let me tinker with option B: advertising all IP addresses with wanton disregard for actually connectivity.

**Comment by delfick** [comment]
where did we get with whether this is viable?

**Comment by Djelibeybi** [comment]
Last I recall, it works. We were disagreeing over network interface selection semantics.

**Comment by Djelibeybi** [comment]
On an only-ever-so-slightly-related but very ironic note, I think bad network interface selection is what's causing most of the issues with the current LIFX integration for Home Assistant.

**Comment by delfick** [comment]
ah yeah. lol.

I thought a problem here was that if interactor and hass were on the same
box they can't both be a zeroconf server or something?

**Comment by Djelibeybi** [comment]
> ah yeah. lol. I thought a problem here was that if interactor and hass were on the same box they can't both be a zeroconf server or something?

No idea where this idea came from. We need Interactor to send mDNS broadcasts so that Home Assistant can see it. There may have been another disagreement over the amount of data we send.

**Comment by delfick** [comment]
I swear I saw something that made this seem unviable. If you say it's fine, then I'll believe you, can always revert if not the case hahah.

This PR seems fine, do you see problems with merging?

**Comment by Djelibeybi** [comment]
I haven't looked at it for a while, but as long as the tests still pass, I think we're good. We could switch it to disabled by default so it's essentially a no-op until enabled?

**Comment by delfick** [comment]
makes sense

**Pull request merged**